src/util/taproot.rs

```diff
@@ -1082,4 +1086,303 @@ mod test {
+            }
+            builder
+        }
+
+        let data = bip_341_read_json();
+        // Check the version of data
+        assert!(data["version"] == 1);
+        let secp = &secp256k1::Secp256k1::verification_only();
+
+        for arr in data["scriptPubKey"].as_array().unwrap() {
+            let internal_key = schnorr::PublicKey::from_str(arr["given"]["internalPubkey"].as_str().unwrap()).unwrap();
+            // process the tree
+            let script_tree = &arr["given"]["scriptTree"];
+            let mut merkle_root = None;
+            if script_tree.is_null() {
+                assert!(arr["intermediary"]["merkleRoot"].is_null());
+            } else {
+                merkle_root = Some(TapBranchHash::from_str(&arr["intermediary"]["merkleRoot"].as_str().unwrap()).unwrap());
+                let leaf_hashes = arr["intermediary"]["leafHashes"].as_array().unwrap();
+                let ctrl_blks = arr["expected"]["scriptPathControlBlocks"].as_array().unwrap();
+                let mut builder = TaprootBuilder::new();
+                let mut leaves = vec![];
+                builder = process_script_trees(&script_tree, builder, &mut leaves, 0);
+                let spend_info = builder.finalize(secp, internal_key).unwrap();
+                for (i, script_ver) in leaves.iter().enumerate() {
+                    let expected_leaf_hash = leaf_hashes[i].as_str().unwrap();
+                    let expected_ctrl_blk = ControlBlock::from_slice(&Vec::<u8>::from_hex(ctrl_blks[i].as_str().unwrap()).unwrap()).unwrap();
+
+                    let leaf_hash = TapLeafHash::from_script(&script_ver.0, script_ver.1);
+                    let ctrl_blk = spend_info.control_block(script_ver).unwrap();
+                    assert_eq!(leaf_hash.to_hex(), expected_leaf_hash);
+                    assert_eq!(ctrl_blk, expected_ctrl_blk);
+                }
+            }
+            let expected_output_key = schnorr::PublicKey::from_str(arr["intermediary"]["tweakedPubkey"].as_str().unwrap()).unwrap();
+            let expected_tweak = TapTweakHash::from_str(arr["intermediary"]["tweak"].as_str().unwrap()).unwrap();
+            let expected_spk = Script::from_str(arr["expected"]["scriptPubKey"].as_str().unwrap()).unwrap();
+            let expected_addr = Address::from_str(arr["expected"]["bip350Address"].as_str().unwrap()).unwrap();
+
+            let tweak = TapTweakHash::from_key_and_tweak(internal_key, merkle_root);
+            let output_key = internal_key.tap_tweak(&secp, merkle_root);
+            let addr = Address::p2tr(&secp, internal_key, merkle_root, Network::Bitcoin);
+            let spk = addr.script_pubkey();
+
+            assert_eq!(expected_output_key, output_key.into_inner());
+            assert_eq!(expected_tweak, tweak);
+            assert_eq!(expected_addr, addr);
+            assert_eq!(expected_spk, spk);
+        }
+    }
+
+    fn bip_341_read_json() -> serde_json::Value {
+        let json_str = r#"{
```

Comment:
did you consider to use the json file directly and use `include_str!` here?

Reply:
Yes, I did consider. However, I was not sure about conventions about where to put test artifacts like the JSON file in the current directory. Any suggestions are appreciated.

Reply:
I like your suggestion, just was not sure about the rust idiomatic way to do things here.